scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala

```diff
@@ -75,8 +76,14 @@ private[dynamictls] object DynamicTLSPeerGroupInternals {
         cause.getMessage: Any,
         ctx.channel().remoteAddress(): Any
       )
-      messageSubject.onNext(UnexpectedError(cause))
+      handleEvent(UnexpectedError(cause))
     }
+
+    private def handleEvent(event: ChannelEvent[M]): Unit =
+      executeSync(messageQueue.tryOffer(event).void)
+
+    private def executeSync(task: Task[Unit]): Unit =
+      task.runSyncUnsafe()
```

Comment:
I wonder, shouldn't we push it on netty event loop of this channel to satisfy single producer scenario ? As i understand when we do push it on this provided scheduler diffrent threads can do the job, there will be always at most one producer, but each time it could be different thread.

Reply:
I'm glad you asked, I hoped to get your input! I tried your [technique](https://github.com/input-output-hk/scalanet/pull/80/files#diff-a967c65483f7c8c1745cf0cbdcc22900R198-R201) first which is to publish on `val eventLoopAsScheduler = monix.execution.Scheduler(nettyChannel.eventLoop())` with `task.runSyncStep(eventLoopAsScheduler)`, however the tests deadlocked, and I thought there were two problems with this:

1. First is that [Task.runSyncStep](https://monix.io/api/3.0/monix/eval/Task.html#runSyncStep(implicits:monix.execution.Scheduler):Either[monix.eval.Task[A],A]) only executes up to the first async boundary, and it returns either the computed result or a task which represents the async calculation. In this case when we just do the `tryOffer` it may fit in one step, I don't know, but the whole stuff that happens in `StaticUDPPeerGroup` didn't. 
2. When I tried `task.runSyncUnsafe(eventLoopAsScheduler, implicitly[CanBlock])` it still deadlocked. I figured the reason was that the `channel.eventLoop` is basically one thread which is currently executing what it does now, which is handling the Netty event. We can't ask it to schedule in our task as well and wait on it, because it gets in the queue of this very thread, it doesn't execute on it straight away, unlike `runSyncStep`. I tried various combinations like executing the data structure maintenance synchronously on the netty scheduler and the replication on the implicit one.

We could definitely execute on the `workGroup` as that one has multiple threads since it's an `EventLoopGroup`, not an `EventLoop`. 

But that would still not answer your question about whether `SPMC` is appropriate and the answer is that I don't really know, I haven't checked under the hood. I don't know if it means that you can only publish from one thread, or whether you cannot publish concurrently.

Reply:
Hm, just read that `queue.clear` should only be called by the consumer, will fix that.

Reply:
Looks like [MpmcArrayQueue.offer](https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpmcArrayQueue.java#L194-L215) executes a loop while [SpmcArrayQueue.offer](https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/SpmcArrayQueue.java#L206-L233) doesn't.

So as long as the checks themselves do a memory fence they should get up-to-date indexes, and since there's no other producer manipulating the queue there shouldn't be a need for a loop.

Reply:
hmm which tests deadlocked ? (I will try to play with it locally a bit)

I am also thinking, what would be the result of just callit `messageQueue.tryOffer(event).runToFuture` (or other scheduler) on channel eventloop. In theory when executing tasks on single threaded scheduler the guaranteed to execute in order of scheduling as underneath single thread scheduler is using linked blocking queue to execute tasks in order. So messages should be delivered in order.
Also on udp peer groups we could probably do that either way as udp in principle does not care about ordering

Reply:
Yeah the `StaticUDPPeerGroup` now starts things in the background because the order doesn't matter, as you said. Let me run it, I can't remember if the deadlock is still present.

Reply:
For example with this change the tests time out:

```scala
  private def executeAsync(task: Task[Unit]): Unit = {
    //task.runAsyncAndForget
    import monix.execution.schedulers.CanBlock
    task.runSyncUnsafe()(Scheduler(serverBinding.channel().eventLoop()), implicitly[CanBlock])
  }
```

```console
$ mill scalanet.test.single io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroupSpec
[54/55] scalanet.test.runMain 
Run starting. Expected test count is: 15
StaticUDPPeerGroupSpec:
StaticUDPPeerGroup
test-14 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup INFO  Server bound to address localhost/127.0.0.1:46237
test-14 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup DEBUG Added UDP client channel from localhost/127.0.0.1:46237 to localhost/127.0.0.1:43195
test-14 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup$ChannelImpl DEBUG Sending client message ByteVector(16777216 bytes, #-648541556)... from localhost/127.0.0.1:46237 to localhost/127.0.0.1:43195
test-14 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup DEBUG Removed UDP client channel from localhost/127.0.0.1:46237 to localhost/127.0.0.1:43195
test-14 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup INFO  Shutting down UDP peer group for peer InetMultiAddress(localhost/127.0.0.1:46237)
- should report an error for sending a message greater than the MTU
test-17 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup INFO  Server bound to address localhost/127.0.0.1:34353
test-14 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup INFO  Server bound to address localhost/127.0.0.1:38099
test-14 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup DEBUG Added UDP client channel from localhost/127.0.0.1:34353 to localhost/127.0.0.1:38099
test-14 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup$ChannelImpl DEBUG Sending client message Hi Bob!... from localhost/127.0.0.1:34353 to localhost/127.0.0.1:38099
nioEventLoopGroup-4-1 io.iohk.scalanet.peergroup.udp.StaticUDPPeerGroup DEBUG Server channel at localhost/127.0.0.1:38099 read message from /127.0.0.1:34353
- should send and receive a message *** FAILED ***
  java.util.concurrent.TimeoutException: Task.runSyncUnsafe(15 seconds)

```

Reply:
Good idea, shall we do this on the TCP groups then?

Reply:
hmm we would have to be sure that messages are provided in correct order, maybe we could first add some test for it ? (also we will probably go through this again when adding back pressoure but this  is future worry)

Reply:
Okay done, works dandy, great stuff :+1: